enterprise/reporting/mediator.go

```diff
@@ -13,77 +13,66 @@ import (
 )
 
 type ReportingMediator struct {
-	log             logger.Logger
-	once            sync.Once
-	enterpriseToken string
-	reporting       types.Reporting
-	errorReporting  types.Reporting
+	log logger.Logger
+
+	delegates []types.Reporting
 }
 
-func (med *ReportingMediator) createReportInstance(backendConfig backendconfig.BackendConfig) types.Reporting {
-	med.log.Debug("Forming reporting instance")
-	reportingEnabled := config.GetBool("Reporting.enabled", types.DefaultReportingEnabled)
-	if !reportingEnabled || med.enterpriseToken == "" {
-		return &NOOP{}
+func NewReportingMediator(ctx context.Context, log logger.Logger, enterpriseToken string, backendConfig backendconfig.BackendConfig) *ReportingMediator {
+	rm := &ReportingMediator{
+		log: log,
 	}
-	reportingHandle := NewFromEnvConfig(med.log)
-	rruntime.Go(func() {
-		reportingHandle.setup(backendConfig)
-	})
-	return reportingHandle
-}
 
-func (med *ReportingMediator) createErrorReportInstance(backendConfig backendconfig.BackendConfig) types.Reporting {
-	med.log.Debug("Forming error reporting instance")
 	reportingEnabled := config.GetBool("Reporting.enabled", types.DefaultReportingEnabled)
-	errorReportingEnabled := config.GetBool("Reporting.errorReporting.enabled", false)
-
-	if !(reportingEnabled && errorReportingEnabled) || med.enterpriseToken == "" {
-		return &NOOP{}
+	if enterpriseToken == "" || !reportingEnabled {
+		return rm
 	}
-	errorReporter := NewEdReporterFromEnvConfig()
+
+	// default reporting implementation
+	defaultReporter := NewDefaultReporter(rm.log)
 	rruntime.Go(func() {
-		errorReporter.setup(backendConfig)
+		defaultReporter.backendConfigSubscriber(ctx, backendConfig)
 	})
-	return errorReporter
-}
+	rm.delegates = append(rm.delegates, defaultReporter)
 
-func NewReportingMediator(log logger.Logger, enterpriseToken string) *ReportingMediator {
-	return &ReportingMediator{
-		log:             log,
-		enterpriseToken: enterpriseToken,
+	// error reporting implementation
+	if config.GetBool("Reporting.errorReporting.enabled", false) {
+		errorReporter := NewErrorDetailReporter()
+		rruntime.Go(func() {
+			errorReporter.backendConfigSubscriber(ctx, backendConfig)
+		})
+		rm.delegates = append(rm.delegates, errorReporter)
```

Comment:
As per the existing logic, should we not create only 1 reporter in case of `Reporting.errorReporting.enabled` and `Reporting.enabled` is true

Reply:
Reporting mediator shall always be created, because even if `Reporting.enabled: false` other reporters might need to be enabled, e.g. the error metadata reporter we will be creating, or a rudder-sources reporter which might come next